lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java

```diff
@@ -179,6 +180,36 @@ boolean isPureDisjunction() {
     return clauses.size() == getClauses(Occur.SHOULD).size() && minimumNumberShouldMatch <= 1;
   }
 
+  /** Whether this query is a two clause disjunction with two term query clauses. */
+  boolean isTwoClausePureDisjunctionWithTerms() {
+    return clauses.size() == 2
+        && isPureDisjunction()
+        && clauses.get(0).getQuery() instanceof TermQuery
+        && clauses.get(1).getQuery() instanceof TermQuery;
+  }
+
+  /**
+   * Rewrite a single two clause disjunction query with terms to two term queries and a conjunction
+   * query using the inclusion–exclusion principle.
+   */
+  Query[] rewriteTwoClauseDisjunctionWithTermsForCount(IndexSearcher indexSearcher)
+      throws IOException {
+    BooleanQuery.Builder newQuery = new BooleanQuery.Builder();
+    Query[] queries = new Query[3];
+    for (int i = 0; i < clauses.size(); i++) {
+      TermQuery termQuery = (TermQuery) clauses.get(i).getQuery();
+      if (termQuery.getTermStates() == null) {
+        termQuery =
+            new TermQuery(
+                termQuery.getTerm(), TermStates.build(indexSearcher, termQuery.getTerm(), true));
```

Comment:
We don't need term statistics here, just a cache for term states.

```suggestion
                termQuery.getTerm(), TermStates.build(indexSearcher, termQuery.getTerm(), false));
```